File: CorporationSalary.java
import java.util.*;

public class CorporationSalary {
  static int n;
  static Map<Integer, List<Integer>> bukas = new HashMap<>();
  static long[] memo;

  public long totalSalary(String[] relations) {
    n = relations.length;
    memo = new long[n];

    for (int i = 0; i < n; i++) {
      List<Integer> list = new ArrayList<>();
      for (int j = 0; j < n; j++) {
        if (relations[i].charAt(j) == 'N')
          continue;

        list.add(j);
      }
      bukas.put(i, list);
    }

    long ans = 0;
    for (int i = 0; i < n; i++) {
      List<Integer> buka = bukas.get(i);
      if (buka.size() == 0) {
        memo[i] = 1;
        ans++;
        continue;
      }

      for (int j = 0; j < buka.size(); j++) {
        ans += salary(buka.get(j));
      }
    }

    return ans;
  }

  public long salary(int i) {
    if (memo[i] > 0) {
      return memo[i];
    }
    List<Integer> buka = bukas.get(i);
    long ret = 0;
    if (buka.size() == 0) {
      ret = 1;
    } else {
      for (int j = 0; j < buka.size(); j++) {
        ret += salary(buka.get(j));
      }
    }
    return memo[i] = ret;
  }
}

File: CorporationSalary.cpp
#include <algorithm>
#include <bitset>
#include <cassert>
#include <cmath>
#include <ctime>
#include <functional>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <tuple>
#include <vector>

#define REP(i,n) for (int i = 0; i < (n); ++i)
#define SORT(a) sort((a).begin(), (a).end());
#define UNIQ(a) SORT(a);(a).erase(unique((a).begin(), (a).end()), (a).end());

using namespace std;
using ll = long long;
using P = pair<int,int>;
using namespace std;

const int MOD = 1000000007; // 10^9 + 7

class CorporationSalary {
  public:
  ll memo[55];
  long long totalSalary(vector <string> relations) {
    REP(i, 55) memo[i] = -1;

    ll ret = 0;

    REP(i, relations.size()) {
      ret += salary(i, relations);
    }

    cout << memo[0] << endl;
    cout << memo[1] << endl;
    cout << memo[2] << endl;
    cout << memo[3] << endl;

    return ret;
  }

  private:
  ll salary(int i, vector<string> relations) {
    if (memo[i] != -1) return memo[i];

    ll ret = 0;
    REP(j, relations.size()) {
      if (relations[i][j] == 'Y') {
        ret += salary(j, relations);
      }
    }

    if (ret == 0) {
      memo[i] = 1;
    } else {
      memo[i] = ret;
    }

    return memo[i];
  }
};

int main() {
  cin.tie(0);
  ios::sync_with_stdio(false);

  CorporationSalary *cs = new CorporationSalary;
  cout << cs->totalSalary({"NNYN","NNYN","NNNN","NYYN"}) << endl;
}